File: registry/src/main/java/it/overzoom/registry/model/Source.java
package it.overzoom.registry.model;

import java.util.Date;
import java.util.List;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

@Document(collection = "source")
public class Source {

    @Id
    private String id;

    private String sn;

    private String phantom;

    private Integer load;

    private Date expirationDate;

    @Indexed
    private String departmentId;

    @Indexed
    private String machineId;

    private Integer completedMeasurements;

    @DBRef
    List<IrradiationCondition> irradiationConditions;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getSn() {
        return sn;
    }

    public void setSn(String sn) {
        this.sn = sn;
    }

    public Date getExpirationDate() {
        return expirationDate;
    }

    public void setExpirationDate(Date expirationDate) {
        this.expirationDate = expirationDate;
    }

    public String getDepartmentId() {
        return departmentId;
    }

    public void setDepartmentId(String departmentId) {
        this.departmentId = departmentId;
    }

    public String getMachineId() {
        return machineId;
    }

    public void setMachineId(String machineId) {
        this.machineId = machineId;
    }

    public List<IrradiationCondition> getIrradiationConditions() {
        return irradiationConditions;
    }

    public void setIrradiationConditions(List<IrradiationCondition> irradiationConditions) {
        this.irradiationConditions = irradiationConditions;
    }

    public Integer getCompletedMeasurements() {
        return completedMeasurements;
    }

    public void setCompletedMeasurements(Integer completedMeasurements) {
        this.completedMeasurements = completedMeasurements;
    }

    public String getPhantom() {
        return phantom;
    }

    public void setPhantom(String phantom) {
        this.phantom = phantom;
    }

    public Integer getLoad() {
        return load;
    }

    public void setLoad(Integer load) {
        this.load = load;
    }

}

File: frontend/src/app/models/Source.ts
import { Machine } from "./Machine";
import { Department } from "./Department";
import { isNaN, isNil, omitBy, overSome } from "lodash-es";
import { IrradiationCondition } from "./IrradiationCondition";

export interface Source {
  id: string,
  sn: string,
  phantom: string,
  load: number,
  expirationDate: string | Date,
  departmentId: string,
  department: Department,
  machineId: number,
  machine: Machine,
  irradiationConditions: IrradiationCondition[],
  completedMeasurements: number
}

export type PartialSource = Partial<Source>;

export type SourceDTO = Source;

export function createSourcePayload(source: any): SourceDTO {
  const sourceDTO = {
    id: source.id,
    sn: source.sn,
    phantom: source.phantom,
    load: source.load,
    expirationDate: source.expirationDate,
    departmentId: source.departmentId,
    department: source.department,
    machineId: source.machineId,
    machine: source.machine,
    completedMeasurements: source.completedMeasurements,
    irradiationConditions: source.irradiationConditions?.filter((i: any) => Object.keys(i).length),
  }
  return <SourceDTO>omitBy(sourceDTO, overSome([isNil, isNaN]));
}